### PR TITLE
finish task1 (comm benchmark) and task2 (llama finetuning on Alpaca). Encountered nan loss.

### PR DESCRIPTION
**Pull request by Edenzzzz**

### CODE REVIEW

File: __pycache__/data_utils.cpython-310.pyc
Comment (ver217) [inline __pycache__/data_utils.cpython-310.pyc:1]: `__pycache__` folder should be ignored

File: src/data_utils.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    
+def preprocess(
+    sources: Sequence[str],
+    targets: Sequence[str],
+    tokenizer: PreTrainedTokenizer,
+) -> Dict:
+    """Preprocess the data by tokenizing."""
+    
+    examples = [s + t for s, t in zip(sources, targets)]
+    if not os.path.exists("alpaca_tokenized.pt"):
```
Comment (ver217) [inline src/data_utils.py]: Dataset path is hard coded. This is not a good idea.

File: test_comm.py
Comment (ver217) [inline test_comm.py:1]: It's so similar to our example. Do you modify anything?

File: tune_llama.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    local_rank = int(os.environ['LOCAL_RANK'])
+    world_size = int(os.environ['WORLD_SIZE'])
+    dist.init_process_group(world_size=world_size, rank=rank,
+                        init_method="env://", backend="nccl")
+    torch.cuda.set_device(local_rank)
+
+@dataclass
+class ModelArguments:
+    model_name_or_path: Optional[str] = field(default="/home/zhongyuting/model/Llama-2-7b-hf", metadata={"help": "Local path to the model or path in huggingface directory"})
+    tp: bool = field(default=False, metadata={"help": "Whether to use tensor parallelism."})
```
Comment (ver217) [inline tune_llama.py]: How do you set tp degree?

File: tune_llama.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+    resize_tokenizer_embedding(
+        new_tokens=special_tokens_dict,
+        tokenizer=tokenizer,
+        model=model,
+    )
+    
+    if model_args.ddp:
+        device = "cuda:" + os.environ['LOCAL_RANK']
+        model = DDP(model, device_ids=[device], output_device=device)
```
Comment (ver217) [inline tune_llama.py]: It cannot use global process group when using tp.

File: test_comm.py
Comment (Edenzzzz) [reply test_comm.py:1]: Yep, added async and broadcast, all-to-all. Efficiently recycling non-patent product

File: src/data_utils.py (hunk 1 above)
Comment (Edenzzzz) [reply src/data_utils.py]: Generally true...but should be fine within a toy program

File: tune_llama.py (hunk 2 above)
Comment (Edenzzzz) [reply tune_llama.py]: Updated

File: tune_llama.py (hunk 3 above)
Comment (Edenzzzz) [reply tune_llama.py]: updated

File: tune_llama.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    dist.init_process_group(world_size=world_size, rank=rank,
+                        init_method="env://", backend="nccl")
+    
+    torch.cuda.set_device(local_rank)
+    
+    # Set up Tensor Parallel within each Data Parallel group
+    tp_size = world_size // dp_size
+    if dp_size > 1:
+        dp_groups = [dist.new_group(ranks=range(group_id * tp_size, (group_id + 1) * tp_size)) for group_id in range(dp_size)]
+        return dp_groups
```
Comment (ver217) [inline tune_llama.py]: Where do you create tp groups?

File: test_tp.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    
+    assert out_tp.shape == out.shape    
+    assert torch.allclose(out, out_tp, atol=1e-5), f"Mean error: {(out - out_tp).mean()}"
+    print("Forward passed!")
+    
+    # Test TP backward
+    out_tp.sum().backward()
+    out.sum().backward()
+
+    column_split_grad = weights_1.grad[size // world_size * rank: size // world_size * (rank + 1)]
```
Comment (ver217) [inline test_tp.py]: You should also check grad of bias

Comment (Edenzzzz) [reply test_tp.py]: Fixed

File: tune_llama.py (hunk 4 above)
Comment (Edenzzzz) [reply tune_llama.py]: Fixed

File: test_tp.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def test_tp_correctness(size):
+    batch_size = 4
+    init_dist(dp_size=1)
+    world_size = dist.get_world_size()
+    rank = dist.get_rank()
+
+    torch.cuda.manual_seed(42)
+    weights_1 = nn.Parameter(torch.ones(size, size, device=f"cuda:{rank}"))
```
Comment (ver217) [inline test_tp.py]: Please use random data

File: test_tp.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    x = torch.ones(batch_size, size, device=f"cuda:{rank}", requires_grad=True)
+    out = F.linear(F.linear(x, weights_1, bias_1), weights_2, bias_2)
+
+    # Check TP forward
+    tp1 = ColumnParallelLinear(weights_1, bias_1)
+    tp2 = RowParallelLinear(weights_2, bias_2)
+    out_tp = tp2(tp1(x))
+    assert out_tp.shape == out.shape
+    assert torch.allclose(out, out_tp, atol=1e-5), f"Mean error: {(out - out_tp).mean()}"
```
Comment (ver217) [inline test_tp.py]: Please use `torch.testing.assert_close()`

File: src/modeling_llama.py [hunk 8]
```diff
@@ -0,0 +1,1779 @@
...
+
+        self.rank = dist.get_rank(group)
+        assert (
+            in_shape % tp_size == 0
+        ), f"The in size must be divisible by the parallel size in TP, but we have {in_shape} / {tp_size}"
+
+        self.group = group
+        self.weight = weight
+        self.bias = (
+            nn.Parameter(bias) if (self.rank == 0 and bias is not None) else None
```
Comment (ver217) [inline src/modeling_llama.py:194]: This leads to memory inbalance.

File: tune_llama.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    dist.init_process_group(world_size=world_size, rank=rank, init_method="env://", backend="nccl")
+    torch.cuda.set_device(local_rank)
+
+    # Set up Tensor Parallel within each Data Parallel group
+    tp_size = world_size // dp_size
+    if dp_size > 1:
+        tp_groups = [
+            dist.new_group(ranks=range(group_id * tp_size, (group_id + 1) * tp_size)) for group_id in range(dp_size)
+        ]
+        return tp_groups
```
Comment (ver217) [inline tune_llama.py]: You should create both tp group and dp group. Only return the group which current process belongs to.

File: tune_llama.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        pbar = tqdm.tqdm(total=total_steps)
+    for epoch in range(training_args.epochs):
+        if sampler:
+            train_loader.sampler.set_epoch(epoch)
+        for step, batch in enumerate(train_loader):
+            batch = to_gpu(batch)
+            with torch.amp.autocast("cuda", dtype=dtype):
+                out = model(**batch)
+                loss = loss_fn(out.logits, batch["labels"]) / training_args.grad_accumulation_steps
+                dist.barrier(group=this_group)
```
Comment (ver217) [inline tune_llama.py]: To need to barrier

File: test_tp.py (hunk 7 above)
Comment (Edenzzzz) [reply test_tp.py]: Fixed

File: src/modeling_llama.py (hunk 8 above)
Comment (Edenzzzz) [reply src/modeling_llama.py:194]: Say dim = 1000, 1k * 4B = 4KB (fp32), which I feel is negligible. Hard to split when the output dim isn't parallelized?

File: tune_llama.py (hunk 9 above)
Comment (Edenzzzz) [reply tune_llama.py]: Cool
